src/LiveComponent/assets/src/url_utils.ts

```diff
@@ -0,0 +1,57 @@
+class AdvancedURLSearchParams extends URLSearchParams {
+    set(name: string, value: any) {
+        if (typeof value !== 'object') {
+            super.set(name, value);
+        } else {
+            this.delete(name);
+            if (Array.isArray(value)) {
+                value.forEach((v) => {
+                    this.append(`${name}[]`, v);
+                });
+            } else {
+                Object.entries(value).forEach(([index, v]) => {
+                    this.append(`${name}[${index}]`, v as string);
+                });
+            }
+        }
+    }
+
+    delete(name: string) {
+        super.delete(name);
+        const pattern = new RegExp(`^${name}(\\[.*])?$`);
+        for (const key of Array.from(this.keys())) {
+            if (key.match(pattern)) {
+                super.delete(key);
+            }
+        }
+    }
+}
```

Comment:
My initial thought is that it looks like you've done a really nice job. If It helps as a reference, I know that Caleb from Livewire worked very hard on the more complex query param creation for Livewire 3 - https://github.com/livewire/livewire/blob/d4839e3b2c23fc71e615e68bc29ff4de95751810/js/plugins/history/index.js#L176

Depending on which more-complex values we handle (e.g. do we handle a `LiveProp` that is an array of strings? What about an array, or array of strings?), we'll need to make sure the query param format works.

Reply:
Thanks for the reference! I'm about to setup more complex use cases to determine the limits of my current implementation. This will help a lot!

Reply:
Can you add a little note, perhaps new the top of the file, with a link to this file in Livewire - something like:

> Adapted from https://github.com/livewire/livewire/blob/d4839e3b2c23fc71e615e68bc29ff4de95751810/js/plugins/history/index.js#L176

It's nice to give Livewire the deserved nod and also will help us, in the future, answer "why did we do it this way?"